Hacked recipe for nektar

pelahi opened the pull request:
Currently it is very difficult to compile nektar when using external
dependencies. However, by altering the recipe to not list dependencies
such as boost, hdf5, blas, nektar will build these libraries internally.
There appears to be no issue with dynamic libraries so nektar should be
fully self-contained.

marcodelapierre commented:
thanks for this work Pascal!

out of curiosity, what types of error occur with external dependencies? do they affect all of blas/lapack, fftw, arpack?

just putting here, for reference, the link to the current source install docs: https://doc.nektar.info/userguide/latest/user-guidese3.html

I noticed an interesting Cmake option, `NEKTAR_ENABLE_SIMD_AVX2`, that might perhaps be good to try out

pelahi commented:
We can git he `NEKTAR_ENABLE_SIMD_AVX2` option an try. Might warrant an update to the recipe so I'll do that too.

There were either issues with cmake configure (could be handled with a patch) or more difficult to debug, compilation issues with the libraries not be exactly how nektar wants them. The former was impacted by external `blas`, `lapack`, `fftw` etc. But the latter was mostly `boost`. It was not obvious exactly what combination of variants would satisfy `nektar`. I thought I had addressed the issues and did manage to reduce the number of compilation bugs but still couldn't get it to zero. It looked like it would have required an update to the boost recipe as well, so not worth it.

marcodelapierre commented:
thanks Pascal!

pelahi commented:
Looks like the avx2 option is not working yet as it introduces compilation errors of third party libraries, specifically in arpack 

```
Determining if the Fortran sgemm exists failed with the following output:
Change Dir: /lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_683e7/fast && gmake[3]: Entering directory '/lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp'
/usr/bin/gmake  -f CMakeFiles/cmTC_683e7.dir/build.make CMakeFiles/cmTC_683e7.dir/build
gmake[4]: Entering directory '/lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp'
Building Fortran object CMakeFiles/cmTC_683e7.dir/testFortranCompiler.f.o
/lus/joey/home/syates/install-trial/spack/lib/spack/env/gcc/gfortran    -c /lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp/testFortranCompiler.f -o CMakeFiles/cmTC_683e7.dir/testFortranCompiler.f.o
Linking Fortran executable cmTC_683e7
/lus/joey/home/pelahi/tmp/cray-sles15-zen2/gcc-10.3.0/cmake-3.21.4-664hnxydbp5y5et6fm43km7rb3lv5nmc/bin/cmake -E cmake_link_script CMakeFiles/cmTC_683e7.dir/link.txt --verbose=1
/lus/joey/home/syates/install-trial/spack/lib/spack/env/gcc/gfortran -dynamic CMakeFiles/cmTC_683e7.dir/testFortranCompiler.f.o -o cmTC_683e7
/usr/bin/ld: CMakeFiles/cmTC_683e7.dir/testFortranCompiler.f.o: in function `main':
testFortranCompiler.f:(.text.startup+0x1b): undefined reference to `sgemm_'
collect2: error: ld returned 1 exit status
gmake[4]: *** [CMakeFiles/cmTC_683e7.dir/build.make:99: cmTC_683e7] Error 1
gmake[4]: Leaving directory '/lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp'
gmake[3]: *** [Makefile:127: cmTC_683e7/fast] Error 2
gmake[3]: Leaving directory '/lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp'



Determining if the Fortran sgemm exists failed with the following output:
Change Dir: /lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_d4352/fast && gmake[3]: Entering directory '/lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp'
/usr/bin/gmake  -f CMakeFiles/cmTC_d4352.dir/build.make CMakeFiles/cmTC_d4352.dir/build
gmake[4]: Entering directory '/lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp'
Building Fortran object CMakeFiles/cmTC_d4352.dir/testFortranCompiler.f.o
/lus/joey/home/syates/install-trial/spack/lib/spack/env/gcc/gfortran    -c /lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp/testFortranCompiler.f -o CMakeFiles/cmTC_d4352.dir/testFortranCompiler.f.o
Linking Fortran executable cmTC_d4352
/lus/joey/home/pelahi/tmp/cray-sles15-zen2/gcc-10.3.0/cmake-3.21.4-664hnxydbp5y5et6fm43km7rb3lv5nmc/bin/cmake -E cmake_link_script CMakeFiles/cmTC_d4352.dir/link.txt --verbose=1
/lus/joey/home/syates/install-trial/spack/lib/spack/env/gcc/gfortran -dynamic CMakeFiles/cmTC_d4352.dir/testFortranCompiler.f.o -o cmTC_d4352  -SSL2
gfortran: error: unrecognized command-line option '-SSL2'
gmake[4]: *** [CMakeFiles/cmTC_d4352.dir/build.make:99: cmTC_d4352] Error 1
gmake[4]: Leaving directory '/lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp'
gmake[3]: *** [Makefile:127: cmTC_d4352/fast] Error 2
gmake[3]: Leaving directory '/lus/joey/home/pelahi/tmp/build_stage/spack-stage-nektar-5.1.0-y7x6uje3dpommkvhhbbgbjgoexk4jesv/spack-build-y7x6uje/ThirdParty/arpack-ng-1.0/CMakeFiles/CMakeTmp'

```

pelahi commented:
For completeness will add the `+avx2` variant anyway as it could be something that might be useful in the future.
